### ISSUE TITLE
Improve Packaging of Edgee WASM Component

### ISSUE THREAD

**Issue by SachaMorard** [edgee-cloud/edgee#128]
## Description

Currently, to build an Edgee WASM component, we need to download wasi_snapshot_preview1.reactor.wasm manually. 

Investigate the current build process and identify potential optimizations.

Explore the possibility of adding wasi_snapshot_preview1.reactor.wasm as a direct dependency in Cargo.toml ?

Implement any feasible optimizations to streamline the build process.

### Additional Context

This improvement will help streamline the build process, making it more efficient and user-friendly, especially to be prepared for Edgee component registry.

## References

Our components repos are:

* [https://github.com/edgee-cloud/ga-component](https://github.com/edgee-cloud/ga-component)
* [https://github.com/edgee-cloud/segment-component](https://github.com/edgee-cloud/segment-component)
* [https://github.com/edgee-cloud/amplitude-component](https://github.com/edgee-cloud/amplitude-component)
* [https://github.com/edgee-cloud/piano-analytics-component](https://github.com/edgee-cloud/piano-analytics-component)
* [https://github.com/edgee-cloud/example-js-component](https://github.com/edgee-cloud/example-js-component)

**Comment by flyaruu** [comment]
I'm not sure what our stance is on Rust nightly? For me nightly seems to work better than stable, at least for WASM.

**Comment by CLEMENTINATOR** [comment]
> Ok, I have played a bit with the amplitude-component. If I use wasm32-wasip2 as target, some slight Cargo.toml wrangling, and Rust nightly, it just works without adaptation.

Can you describe the changes needed in Cargo.toml ? 
Otherwise, regarding nightly, (and wasip2), I guess as long as stuff doesn't break, lets go for it!

**Comment by Jer1605** [comment]
### Question @SachaMorard 

Why was the decision made initially to download `wasi_snapshot_preview1.reactor.wasm` manually instead of including it directly as a dependency in `Cargo.toml`? 

Was there a specific limitation or compatibility issue that led to this approach? Understanding the original reasoning might help us avoid potential pitfalls.

**Comment by KokaKiwi** [comment]
I would suggest taking a loot at wasmtime's [`wasi-preview1-component-adapter-provider`](https://crates.io/crates/wasi-preview1-component-adapter-provider) crate, we could integrate that into the build process (like using Rust `build.rs` "build script")
And maybe later get the whole component building into a dedicated tool (we could integrate later into the future `edgee` CLI), so we wouldn't have to ask component developers to "manually" write the `wasm-tools` command

**Comment by SachaMorard** [comment]
> ### Question @SachaMorard
> Why was the decision made initially to download `wasi_snapshot_preview1.reactor.wasm` manually instead of including it directly as a dependency in `Cargo.toml`?
> 
> Was there a specific limitation or compatibility issue that led to this approach? Understanding the original reasoning might help us avoid potential pitfalls.

Unfortunately, I don't know. I'm not the one who built the first component ;)

**Comment by SachaMorard** [comment]
> I would suggest taking a loot at wasmtime's [`wasi-preview1-component-adapter-provider`](https://crates.io/crates/wasi-preview1-component-adapter-provider) crate, we could integrate that into the build process (like using Rust `build.rs` "build script"):
> 
> ```rust
> // build.rs
> use wasi_preview1_component_adapter_provider::WASI_SNAPSHOT_PREVIEW1_REACTOR_ADAPTER;
> 
> fn main() {
>     std::fs::write(
>         "wasi_snapshot_preview1.reactor.wasm",
>         WASI_SNAPSHOT_PREVIEW1_REACTOR_ADAPTER
>     ).expect("failed to write bytes to file");
> }
> ```
> 
> And maybe later get the whole component building into a dedicated tool (and integrate later into the future `edgee` CLI?), so we wouldn't have to ask component developers to "manually" write the `wasm-tools` command

Ok for the rust components, but what about the js/python/go/... components?

**Comment by SachaMorard** [comment]
> I'm not sure what our stance is on Rust nightly? For me nightly seems to work better than stable, at least for WASM.

I assume that using Rust nightly to build a component requires the same Rust version for the proxy, right?

**Comment by flyaruu** [comment]
I don't think so, the component doesn't even need to be based on Rust at all.

**Comment by SachaMorard** [comment]
Even with something like `./edgee component build` ? ;)

**Comment by KokaKiwi** [comment]
The Rust version used to build a component is completely uncorrelated to the Rust version used to build Edgee CLI even when using `edgee component build` since there's no interaction at ABI-level between the two, the only interfacing done is with WASM Component which has its own ABI, so you could build a Rust component using Rust nightly and another one with Rust 1.80.0 if you really want to and it would still works :)

**Comment by alexcasalboni** [comment]
From a developer experience (DX) point of view, it would be great to **NOT** having to install Rust locally to build and run a new component in Go, JS, Python, etc.

Ideally, you just `brew install edgee` and then `edgee component create/build/run` without having to worry about Rust or `wasi_snapshot_preview1.reactor.wasm`.

Would it be possible to include the .wasm file directly in the `edgee` binary? it's only 54KB, right?

**Comment by flyaruu** [comment]
So some updates:
- The issue with the url crate (requiring 'default-features=false', is also resolved by using the latest version (2.5.3 instead of 2.5.2), so it looks like it was a bug in url
- I think the issue with the stable compiler (or rather the linker) is also a bug which has been fixed in nightly. It is also fixed in the beta channel, which, if nothing weird happens, will be stable next Thursday.

So in conclusion: I think the discussion about nightly will be irrelevant very soon.
